anndata/utils.py

```diff
@@ -60,17 +64,37 @@ def make_index_unique(index: pd.Index, join: str = "-"):
     """
     if index.is_unique:
         return index
-    from collections import defaultdict
+    from collections import Counter
 
-    values = index.values
+    values = index.values.copy()
     indices_dup = index.duplicated(keep="first")
     values_dup = values[indices_dup]
-    counter = defaultdict(lambda: 0)
+    values_set = set(values)
+    counter = Counter()
+    issue_interpretation_warning = False
+    example_colliding_values = []
     for i, v in enumerate(values_dup):
-        counter[v] += 1
-        values_dup[i] += join + str(counter[v])
+        while True:
+            counter[v] += 1
+            tentative_new_name = v + join + str(counter[v])
+            if tentative_new_name not in values_set:
+                values_set.add(tentative_new_name)
+                values_dup[i] = tentative_new_name
+                break
+            issue_interpretation_warning = True
+            if len(example_colliding_values) < 5:
+                example_colliding_values.append(tentative_new_name)
+
+    if issue_interpretation_warning:
+        warnings.warn(
+            f"Suffix used ({join}[0-9]+) to deduplicate index values may make index "
+            + "values difficult to interpret. There values with a similar suffixes in "
+            + "the index. Consider using a different join parameter."
```

Comment:
I like our errors and warnings to provide as explicit of a solution as possible. Could this be more like:

```suggestion
            + "the index. Consider using a different delimiter by passing `join={delimiter}`."
```